File: src/Web/src/app/core/banking/banking.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface BankAccountDto {
  id: string;
  customerId: string;
  iban: string;
  accountNumber: string;
  type: string;
  status: string;
  balanceAmount: number;
  currency: string;
  createdAt: string;
}

export interface TransactionDto {
  id: string;
  accountId: string;
  type: string;
  amount: number;
  currency: string;
  description: string;
  referenceNumber: string;
  transactionDate: string;
}

export interface CreateAccountCommand {
  customerId: string;
  type: string;
  initialDeposit: number;
  currency: string;
}

export interface DepositCommand {
  amount: number;
  currency: string;
  description: string;
}

export interface WithdrawCommand {
  amount: number;
  currency: string;
  description: string;
}

@Injectable({
  providedIn: 'root'
})
export class BankingService {
  private http = inject(HttpClient);

  private get API_URL(): string {
    if (typeof window !== 'undefined' && window.location.hostname !== 'localhost' && window.location.hostname !== '127.0.0.1') {
      return 'https://finhub-w9xq.onrender.com/api/v1/accounts';
    }
    return 'http://localhost:5000/api/v1/accounts';
  }

  public getAccountsByCustomer(customerId: string): Observable<BankAccountDto[]> {
    return this.http.get<BankAccountDto[]>(`${this.API_URL}/customer/${customerId}`);
  }

  public getAccountById(accountId: string): Observable<BankAccountDto> {
    return this.http.get<BankAccountDto>(`${this.API_URL}/${accountId}`);
  }

  public createAccount(command: CreateAccountCommand): Observable<BankAccountDto> {
    return this.http.post<BankAccountDto>(this.API_URL, command);
  }

  public deposit(accountId: string, command: DepositCommand): Observable<BankAccountDto> {
    return this.http.post<BankAccountDto>(`${this.API_URL}/${accountId}/deposit`, command);
  }

  public withdraw(accountId: string, command: WithdrawCommand): Observable<BankAccountDto> {
    return this.http.post<BankAccountDto>(`${this.API_URL}/${accountId}/withdraw`, command);
  }

  public getTransactions(accountId: string): Observable<TransactionDto[]> {
    return this.http.get<TransactionDto[]>(`${this.API_URL}/${accountId}/transactions`);
  }
}

File: src/FinHub/FinHub.Application/Modules/Banking/Services/BankingService.cs
using FinHub.Application.Common.Interfaces;
using FinHub.Application.Modules.Banking.DTOs;
using FinHub.Domain.Entities;
using FinHub.Domain.Enums;
using FinHub.Domain.ValueObjects;

namespace FinHub.Application.Modules.Banking.Services;

public sealed class BankingService : IBankingService
{
    private readonly IBankAccountRepository _accountRepository;
    private readonly ICustomerRepository _customerRepository;

    public BankingService(IBankAccountRepository accountRepository, ICustomerRepository customerRepository)
    {
        _accountRepository = accountRepository;
        _customerRepository = customerRepository;
    }

    public async Task<BankAccountDto> CreateAccountAsync(CreateAccountCommand command, CancellationToken cancellationToken = default)
    {
        var customer = await _customerRepository.GetByIdAsync(command.CustomerId, cancellationToken)
            ?? throw new KeyNotFoundException($"Customer with ID '{command.CustomerId}' was not found.");

        if (!Enum.TryParse<AccountType>(command.Type, true, out var type))
        {
            throw new ArgumentException($"Invalid account type '{command.Type}'. Valid types: Savings, Current, Investment, Credit.");
        }

        if (!Enum.TryParse<Currency>(command.Currency, true, out var currency))
        {
            throw new ArgumentException($"Invalid currency '{command.Currency}'. Valid currencies: SAR, USD, EUR, GBP, AED.");
        }

        var initialBalance = command.InitialDeposit > 0 ? new Money(command.InitialDeposit, currency) : null;
        var account = BankAccount.Create(customer.Id, type, initialBalance);

        await _accountRepository.AddAsync(account, cancellationToken);

        return MapToDto(account);
    }

    public async Task<BankAccountDto> GetAccountByIdAsync(Guid accountId, CancellationToken cancellationToken = default)
    {
        var account = await _accountRepository.GetByIdAsync(accountId, cancellationToken)
            ?? throw new KeyNotFoundException($"Bank account with ID '{accountId}' was not found.");

        return MapToDto(account);
    }

    public async Task<List<BankAccountDto>> GetAccountsByCustomerIdAsync(Guid customerId, CancellationToken cancellationToken = default)
    {
        var accounts = await _accountRepository.GetByCustomerIdAsync(customerId, cancellationToken);
        return accounts.Select(MapToDto).ToList();
    }

    public async Task<BankAccountDto> DepositAsync(Guid accountId, DepositCommand command, CancellationToken cancellationToken = default)
    {
        var account = await _accountRepository.GetByIdAsync(accountId, cancellationToken)
            ?? throw new KeyNotFoundException($"Bank account with ID '{accountId}' was not found.");

        if (!Enum.TryParse<Currency>(command.Currency, true, out var currency))
        {
            throw new ArgumentException($"Invalid currency '{command.Currency}'.");
        }

        var depositMoney = new Money(command.Amount, currency);
        account.Deposit(depositMoney, command.Description);

        await _accountRepository.UpdateAsync(account, cancellationToken);

        return MapToDto(account);
    }

    public async Task<BankAccountDto> WithdrawAsync(Guid accountId, WithdrawCommand command, CancellationToken cancellationToken = default)
    {
        var account = await _accountRepository.GetByIdAsync(accountId, cancellationToken)
            ?? throw new KeyNotFoundException($"Bank account with ID '{accountId}' was not found.");

        if (!Enum.TryParse<Currency>(command.Currency, true, out var currency))
        {
            throw new ArgumentException($"Invalid currency '{command.Currency}'.");
        }

        var withdrawMoney = new Money(command.Amount, currency);
        account.Withdraw(withdrawMoney, command.Description);

        await _accountRepository.UpdateAsync(account, cancellationToken);

        return MapToDto(account);
    }

    public async Task<List<TransactionDto>> GetTransactionsAsync(Guid accountId, CancellationToken cancellationToken = default)
    {
        var account = await _accountRepository.GetByIdAsync(accountId, cancellationToken)
            ?? throw new KeyNotFoundException($"Bank account with ID '{accountId}' was not found.");

        return account.Transactions
            .OrderByDescending(t => t.TransactionDate)
            .Select(t => new TransactionDto
            {
                Id = t.Id,
                AccountId = t.AccountId,
                Type = t.Type.ToString(),
                Amount = t.Amount.Amount,
                Currency = t.Amount.Currency.ToString(),
                Description = t.Description,
                ReferenceNumber = t.ReferenceNumber,
                TransactionDate = t.TransactionDate
            })
            .ToList();
    }

    public async Task<List<TransactionDto>> GetCustomerTransactionsAsync(Guid customerId, int page = 1, int pageSize = 25, CancellationToken cancellationToken = default)
    {
        if (page < 1)
            page = 1;

        if (pageSize < 1)
            pageSize = 25;

        var customer = await _customerRepository.GetByIdAsync(customerId, cancellationToken)
            ?? throw new KeyNotFoundException($"Customer with ID '{customerId}' was not found.");

        var accounts = await _accountRepository.GetByCustomerIdAsync(customer.Id, cancellationToken);
        var transactions = accounts
            .SelectMany(a => a.Transactions)
            .OrderByDescending(t => t.TransactionDate)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .Select(t => new TransactionDto
            {
                Id = t.Id,
                AccountId = t.AccountId,
                Type = t.Type.ToString(),
                Amount = t.Amount.Amount,
                Currency = t.Amount.Currency.ToString(),
                Description = t.Description,
                ReferenceNumber = t.ReferenceNumber,
                TransactionDate = t.TransactionDate
            })
            .ToList();

        return transactions;
    }

    private static BankAccountDto MapToDto(BankAccount account)
    {
        return new BankAccountDto
        {
            Id = account.Id,
            CustomerId = account.CustomerId,
            Iban = account.Iban.Value,
            AccountNumber = account.AccountNumber,
            Type = account.Type.ToString(),
            Status = account.Status.ToString(),
            BalanceAmount = account.Balance.Amount,
            Currency = account.Balance.Currency.ToString(),
            CreatedAt = account.CreatedAt
        };
    }
}
